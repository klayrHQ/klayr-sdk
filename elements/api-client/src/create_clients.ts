/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { APIClient } from './api_client';
import { IPCChannel } from './ipc_channel';
import { WSChannel } from './ws_channel';

import { Channel } from './types';

export const createClient = async (
	channel: Omit<Channel, 'connect' | 'disconnect'>,
): Promise<APIClient> => {
	const client = new APIClient(channel as Channel);
	await client.init();

	return client;
};

export const createIPCClient = async (dataPath: string): Promise<APIClient> => {
	const ipcChannel = new IPCChannel(dataPath);
	await ipcChannel.connect();

	return createClient(ipcChannel);
};

export const createWSClient = async (url: string): Promise<APIClient> => {
	const wsChannel = new WSChannel(url);
	await wsChannel.connect();

	return createClient(wsChannel);
};
