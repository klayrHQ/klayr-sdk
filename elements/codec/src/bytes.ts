/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { writeUInt32, readUInt32 } from './varint';

export const writeBytes = (bytes: Buffer): Buffer =>
	Buffer.concat([writeUInt32(bytes.length), bytes]);

export const readBytes = (buffer: Buffer, offset: number): [Buffer, number] => {
	const [byteLength, keySize] = readUInt32(buffer, offset);
	return [buffer.subarray(offset + keySize, offset + keySize + byteLength), byteLength + keySize];
};
