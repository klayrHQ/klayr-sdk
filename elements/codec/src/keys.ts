/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-bitwise */
/* eslint-disable @typescript-eslint/no-unsafe-return */
// https://developers.google.com/protocol-buffers/docs/encoding#structure
const WIRE_TYPE_TWO = 2; // string, bytes, object, array
const WIRE_TYPE_ZERO = 0; // uint32, uint64, sint32, sint64, boolean

export const readKey = (value: number): [number, number] => {
	const wireType = value & 7;
	if (wireType === WIRE_TYPE_TWO || wireType === WIRE_TYPE_ZERO) {
		const fieldNumber = value >>> 3;

		return [fieldNumber, wireType];
	}

	throw new Error('Value yields unsupported wireType');
};
