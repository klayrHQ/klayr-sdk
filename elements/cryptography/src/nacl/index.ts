/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
// eslint-disable-next-line import/no-cycle
import { NaclInterface } from './nacl_types';

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires
let lib: NaclInterface = require('./slow');

// Use try/catch for browser fallback support
try {
	if (process.env.NACL_FAST !== 'disable') {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-require-imports,global-require
		lib = require('./fast');
	}
	// eslint-disable-next-line no-empty
} catch (err) {}

export const NACL_SIGN_PUBLICKEY_LENGTH = 32;

export const NACL_SIGN_SIGNATURE_LENGTH = 64;

export const {
	box,
	openBox,
	signDetached,
	verifyDetached,
	getRandomBytes,
	getKeyPair,
	getPublicKey,
} = lib;
