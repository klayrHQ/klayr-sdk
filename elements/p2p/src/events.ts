/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
// P2P general
export const EVENT_NEW_INBOUND_PEER = 'EVENT_NEW_INBOUND_PEER';
export const EVENT_NEW_INBOUND_PEER_CONNECTION = 'EVENT_NEW_INBOUND_PEER_CONNECTION';
export const EVENT_FAILED_TO_ADD_INBOUND_PEER = 'EVENT_FAILED_TO_ADD_INBOUND_PEER';
export const EVENT_NETWORK_READY = 'EVENT_NETWORK_READY';

// P2P pool
export const EVENT_REMOVE_PEER = 'EVENT_REMOVE_PEER';

// Peer Local emitted events
export const EVENT_REQUEST_RECEIVED = 'EVENT_REQUEST_RECEIVED';
export const EVENT_INVALID_REQUEST_RECEIVED = 'EVENT_INVALID_REQUEST_RECEIVED';
export const EVENT_MESSAGE_RECEIVED = 'EVENT_MESSAGE_RECEIVED';
export const EVENT_INVALID_MESSAGE_RECEIVED = 'EVENT_INVALID_MESSAGE_RECEIVED';
export const EVENT_BAN_PEER = 'EVENT_BAN_PEER';
export const EVENT_DISCOVERED_PEER = 'EVENT_DISCOVERED_PEER';
export const EVENT_UPDATED_PEER_INFO = 'EVENT_UPDATED_PEER_INFO';
export const EVENT_FAILED_PEER_INFO_UPDATE = 'EVENT_FAILED_PEER_INFO_UPDATE';
export const EVENT_FAILED_TO_COLLECT_PEER_DETAILS_ON_CONNECT =
	'EVENT_FAILED_TO_COLLECT_PEER_DETAILS_ON_CONNECT';
export const EVENT_FAILED_TO_FETCH_PEERS = 'EVENT_FAILED_TO_FETCH_PEERS';
export const EVENT_FAILED_TO_FETCH_PEER_INFO = 'EVENT_FAILED_TO_FETCH_PEER_INFO';
export const EVENT_FAILED_TO_PUSH_NODE_INFO = 'EVENT_FAILED_TO_PUSH_NODE_INFO';
export const EVENT_FAILED_TO_SEND_MESSAGE = 'EVENT_FAILED_TO_SEND_MESSAGE';

// Peer base
export const REMOTE_SC_EVENT_RPC_REQUEST = 'rpc-request';
export const REMOTE_SC_EVENT_MESSAGE = 'remote-message';

// P2P Protocol messages
export const REMOTE_EVENT_POST_NODE_INFO = 'postNodeInfo';
export const REMOTE_EVENT_RPC_GET_NODE_INFO = 'getNodeInfo';
export const REMOTE_EVENT_RPC_GET_PEERS_LIST = 'getPeers';
export const REMOTE_EVENT_PING = 'ping';
export const REMOTE_EVENT_PONG = 'pong';

// P2P Protocol messages list
export const PROTOCOL_EVENTS_TO_RATE_LIMIT: Set<string> = new Set([
	REMOTE_EVENT_RPC_GET_NODE_INFO,
	REMOTE_EVENT_RPC_GET_PEERS_LIST,
]);

// Inbound peer
export const EVENT_CLOSE_INBOUND = 'EVENT_CLOSE_INBOUND';
export const EVENT_INBOUND_SOCKET_ERROR = 'EVENT_INBOUND_SOCKET_ERROR';

// Outbound peer
export const EVENT_CONNECT_OUTBOUND = 'EVENT_CONNECT_OUTBOUND';
export const EVENT_CONNECT_ABORT_OUTBOUND = 'EVENT_CONNECT_ABORT_OUTBOUND';
export const EVENT_CLOSE_OUTBOUND = 'EVENT_CLOSE_OUTBOUND';
export const EVENT_OUTBOUND_SOCKET_ERROR = 'EVENT_OUTBOUND_SOCKET_ERROR';
