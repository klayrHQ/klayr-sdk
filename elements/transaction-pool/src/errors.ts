/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export class TransactionPoolError extends Error {
	public message: string;
	public id: Buffer;
	public dataPath: string;
	public actual?: string | number;
	public expected?: string | number;
	public constructor(
		message = '',
		id = Buffer.alloc(0),
		dataPath = '',
		actual?: string | number,
		expected?: string | number,
	) {
		super();
		this.message = message;
		this.id = id;
		this.name = 'TransactionPoolError';
		this.dataPath = dataPath;
		this.actual = actual;
		this.expected = expected;
	}

	public toString(): string {
		const defaultMessage = `TransactionPool: ${this.id.toString('hex')} failed to process at ${
			this.dataPath
		}: ${this.message}`;
		const withActual = this.actual
			? // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
			  `${defaultMessage}, actual: ${this.actual}`
			: defaultMessage;
		const withExpected = this.expected
			? // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
			  `${withActual}, expected: ${this.expected}`
			: withActual;

		return withExpected;
	}
}
