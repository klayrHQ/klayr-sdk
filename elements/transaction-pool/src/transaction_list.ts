/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { dataStructures } from '@klayr/utils';
import { Transaction } from './types';

export interface TransactionListOptions {
	readonly maxSize?: number;
	readonly minReplacementFeeDifference?: bigint;
}

const DEFAULT_MAX_SIZE = 64;
export const DEFAULT_MINIMUM_REPLACEMENT_FEE_DIFFERENCE = BigInt(10);

type AddStatus =
	| { added: true; removedID?: Buffer; reason?: undefined }
	| { added: false; removedID?: Buffer; reason: string };

export class TransactionList {
	public readonly address: Buffer;

	private _processable: Array<bigint>;
	private readonly _transactions: { [nonce: string]: Transaction };
	// Value is not needed here because it is stored separately in the _transactions
	private readonly _nonceHeap: dataStructures.MinHeap<undefined, bigint>;
	private readonly _maxSize: number;
	private readonly _minReplacementFeeDifference: bigint;

	public constructor(address: Buffer, options?: TransactionListOptions) {
		this.address = address;
		this._transactions = {};
		this._nonceHeap = new dataStructures.MinHeap<undefined, bigint>();
		this._processable = [];
		this._maxSize = options?.maxSize ?? DEFAULT_MAX_SIZE;
		this._minReplacementFeeDifference =
			options?.minReplacementFeeDifference ?? DEFAULT_MINIMUM_REPLACEMENT_FEE_DIFFERENCE;
	}

	public get(nonce: bigint): Transaction | undefined {
		return this._transactions[nonce.toString()];
	}

	public add(incomingTx: Transaction, processable = false): AddStatus {
		const existingTx = this._transactions[incomingTx.nonce.toString()];
		// If the same nonce already exist in the pool try to replace
		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
		if (existingTx) {
			// If the fee is lower than the original fee + replacement, reject
			if (incomingTx.fee < existingTx.fee + this._minReplacementFeeDifference) {
				return {
					added: false,
					reason: 'Incoming transaction fee is not sufficient to replace existing transaction',
				};
			}
			// Mark this and all subsequent nonce unprocessable
			this._demoteAfter(incomingTx.nonce);
			this._transactions[incomingTx.nonce.toString()] = incomingTx;

			return { added: true, removedID: existingTx.id };
		}

		const highestNonce = this._highestNonce();
		let removedID;
		if (this._nonceHeap.count >= this._maxSize) {
			// If incoming nonce is bigger than the highest nonce, then reject
			if (incomingTx.nonce > highestNonce) {
				return {
					added: false,
					reason: 'Incoming transaction exceeds maximum transaction limit per account',
				};
			}
			// If incoming nonce is lower than the highest nonce, remove the largest nonce transaction instead
			removedID = this.remove(highestNonce);
		}

		this._transactions[incomingTx.nonce.toString()] = incomingTx;
		this._nonceHeap.push(incomingTx.nonce, undefined);
		// If this transaction is processable and it is the first one in the list
		if (processable && this._processable.length === 0) {
			this._processable.push(incomingTx.nonce);
		}

		return { added: true, removedID };
	}

	public remove(nonce: bigint): Buffer | undefined {
		const removingTx = this._transactions[nonce.toString()];
		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
		if (!removingTx) {
			return undefined;
		}
		delete this._transactions[nonce.toString()];
		// Recreate heap: it could remove in the middle of the heap
		const { keys } = this._nonceHeap;
		this._nonceHeap.clear();
		for (const key of keys) {
			if (key !== nonce) {
				this._nonceHeap.push(key, undefined);
			}
		}
		this._demoteAfter(nonce);

		return removingTx.id;
	}

	public promote(txs: ReadonlyArray<Transaction>): boolean {
		// Promote if only all ID are still existing and the same
		const promotingNonces = [];
		for (const tx of txs) {
			const promotingTx = this._transactions[tx.nonce.toString()];
			// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
			if (!promotingTx) {
				return false;
			}
			if (tx.id !== promotingTx.id) {
				return false;
			}
			promotingNonces.push(tx.nonce);
		}
		this._processable = Array.from(new Set([...this._processable, ...promotingNonces]));
		this._sortProcessable();

		return true;
	}

	public get size(): number {
		return this._nonceHeap.count;
	}

	public getProcessable(): ReadonlyArray<Transaction> {
		const txs = [];
		for (const nonce of this._processable) {
			txs.push(this._transactions[nonce.toString()]);
		}

		return txs;
	}

	public getUnprocessable(): ReadonlyArray<Transaction> {
		if (this._nonceHeap.count === 0) {
			return [];
		}
		if (this._processable.length === this._nonceHeap.count) {
			return [];
		}
		const clonedHeap = this._nonceHeap.clone();
		// Clone heap with the root node as the first unprocessable transaction
		for (const _ of this._processable) {
			clonedHeap.pop();
		}
		const remainingCount = clonedHeap.count;
		const unprocessableTx: Transaction[] = [];
		for (let i = 0; i < remainingCount; i += 1) {
			const { key } = clonedHeap.pop() as { key: bigint };
			unprocessableTx.push(this._transactions[key.toString()]);
		}

		return unprocessableTx;
	}

	public getPromotable(): ReadonlyArray<Transaction> {
		if (this._nonceHeap.count === 0) {
			return [];
		}
		if (this._processable.length === this._nonceHeap.count) {
			return [];
		}
		const clonedHeap = this._nonceHeap.clone();
		// Make cloned heap root to unprocessable
		for (const _ of this._processable) {
			clonedHeap.pop();
		}
		const firstUnprocessable = clonedHeap.pop();
		if (!firstUnprocessable) {
			return [];
		}
		if (this._processable.length !== 0) {
			const highestProcessableNonce = this._processable[this._processable.length - 1];
			if (firstUnprocessable.key !== highestProcessableNonce + BigInt(1)) {
				return [];
			}
		}
		const promotableTx = [this._transactions[firstUnprocessable.key.toString()]];

		const remainingNonces = clonedHeap.count;
		let lastPromotedNonce = this._transactions[firstUnprocessable.key.toString()].nonce;
		for (let i = 0; i < remainingNonces; i += 1) {
			const { key } = clonedHeap.pop() as { key: bigint };
			if (lastPromotedNonce + BigInt(1) === key) {
				promotableTx.push(this._transactions[key.toString()]);
				lastPromotedNonce = this._transactions[key.toString()].nonce;
			}
		}

		return promotableTx;
	}

	private _demoteAfter(nonce: bigint): void {
		this._processable = this._processable.filter(processableNonce => processableNonce < nonce);
		this._sortProcessable();
	}

	private _highestNonce(): bigint {
		const highestNonce = BigInt(-1);
		const { keys } = this._nonceHeap;
		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
		if (!keys) {
			return highestNonce;
		}

		return keys.reduce((prev, current) => {
			if (current > prev) {
				return current;
			}

			return prev;
		}, highestNonce);
	}

	private _sortProcessable(): void {
		this._processable.sort((a, b) => {
			if (a - b > BigInt(0)) {
				return 1;
			}
			if (a - b < BigInt(0)) {
				return -1;
			}

			return 0;
		});
	}
}
