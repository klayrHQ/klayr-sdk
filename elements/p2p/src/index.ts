/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as constants from './constants';
import * as errors from './errors';
import * as events from './events';
import { P2P } from './p2p';
import * as p2pRequest from './p2p_request';
import * as p2pTypes from './types';
import * as p2pSchemas from './schema';

export { constants, errors, events, p2pRequest, p2pTypes, p2pSchemas, P2P };
