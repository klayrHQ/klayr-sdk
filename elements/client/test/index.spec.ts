/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
// eslint-disable-next-line import/namespace
import * as client from '../src';

describe('klayr-client', () => {
	it('should not export unexpected libraries', () => {
		return expect(Object.keys(client)).toMatchSnapshot();
	});
});
