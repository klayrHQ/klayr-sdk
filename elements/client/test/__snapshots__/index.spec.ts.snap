// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`klayr-client should not export unexpected libraries 1`] = `
[
  "apiClient",
  "cryptography",
  "passphrase",
  "transactions",
  "utils",
  "tree",
  "validator",
  "codec",
]
`;
